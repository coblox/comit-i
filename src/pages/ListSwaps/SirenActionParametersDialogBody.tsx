import {
  Button,
  DialogActions,
  DialogContent,
  DialogTitle
} from "@material-ui/core";
import { ButtonProps } from "@material-ui/core/Button";
import { DialogTitleProps } from "@material-ui/core/DialogTitle";
import React, { useReducer } from "react";
import { Action, Field } from "../../../gen/siren";
import EthereumAddressTextField from "../../components/EthereumAddressTextField";
import TextField from "../../components/TextField";

interface Props {
  action: Action;
  onClose: () => void;
  onSubmit: (parameters: any) => void;
}

/*
 * A component to be placed into a Dialog to gather all necessary parameters from the user for executing the given action.
 */
export default function SirenActionParametersDialogBody({
  action,
  onClose,
  onSubmit
}: Props) {
  const method = action.method || "GET";
  const title = action.title || action.name;

  const [actionPayload, dispatch] = useReducer(actionPayloadReducer, {});

  const formFields = action.fields
    ? action.fields.map(field =>
        renderField(field, actionPayload[field.name] || "", newValue =>
          dispatch({
            type: "onFieldChanged",
            payload: {
              fieldName: field.name,
              newValue
            }
          })
        )
      )
    : [];

  return (
    <React.Fragment>
      <Title method={method} title={title} />
      <form
        onSubmit={event => {
          event.preventDefault();
          onSubmit(actionPayload);
        }}
      >
        <DialogContent>{formFields}</DialogContent>
        <DialogActions>
          <PrimaryActionButton
            method={method}
            title={title}
            type={"submit"}
            data-cy={`${action.name}-button`}
          />
          <Button
            type={"button"}
            variant={"contained"}
            color={"secondary"}
            onClick={onClose}
            data-cy={"close-button"}
          >
            Close
          </Button>
        </DialogActions>
      </form>
    </React.Fragment>
  );
}

function renderField(
  field: Field,
  value: any,
  onChange: (newValue: any) => void
) {
  if (field.class.includes("ethereum") && field.class.includes("address")) {
    return (
      <EthereumAddressTextField
        data-cy={"action-text-field"}
        required={true}
        key={field.name}
        label={field.title}
        onAddress={onChange}
        onChange={event => onChange(event.target.value)}
        value={value}
      />
    );
  }

  return (
    <TextField
      data-cy={"action-text-field"}
      required={true}
      key={field.name}
      label={field.title}
      type={field.type}
      value={value}
      onChange={event => onChange(event.target.value)}
    />
  );
}

interface ActionPayload {
  [name: string]: any;
}

// tslint:disable-next-line:interface-over-type-literal
type ReducerAction = {
  type: "onFieldChanged";
  payload: {
    fieldName: string;
    newValue: any;
  };
};

function actionPayloadReducer(state: ActionPayload, action: ReducerAction) {
  switch (action.type) {
    case "onFieldChanged": {
      return {
        ...state,
        [action.payload.fieldName]: action.payload.newValue
      };
    }
  }

  return {};
}

interface DisplayProps {
  method: string;
  title: string;
}

function Title({
  method,
  title,
  ...dialogTitleProps
}: DisplayProps & DialogTitleProps) {
  switch (method) {
    case "GET":
      return (
        <DialogTitle
          {...dialogTitleProps}
        >{`To fetch the '${title}' action, please fill these fields.`}</DialogTitle>
      );
    default:
      return (
        <DialogTitle
          {...dialogTitleProps}
        >{`To execute the '${title}' action, please fill these fields.`}</DialogTitle>
      );
  }
}

function PrimaryActionButton({
  method,
  title,
  ...buttonProps
}: DisplayProps & ButtonProps) {
  switch (method) {
    case "GET":
      return (
        <Button variant={"contained"} color={"primary"} {...buttonProps}>
          {`Fetch ${title}`}
        </Button>
      );
    default:
      return (
        <Button variant={"contained"} color={"primary"} {...buttonProps}>
          {`Execute ${title}`}
        </Button>
      );
  }
}
