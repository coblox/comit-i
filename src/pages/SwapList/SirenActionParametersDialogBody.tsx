import {
  Button,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle
} from "@material-ui/core";
import { DialogTitleProps } from "@material-ui/core/DialogTitle";
import React, { useReducer } from "react";
import { Action, Field } from "../../../gen/siren";
import BitcoinNanoLedgerAddressTextField from "../../components/BitcoinNanoLedgerAddressTextField";
import EthereumAddressTextField from "../../components/EthereumAddressTextField";
import TextField from "../../components/TextField";

interface Props {
  action: Action;
  onClose: () => void;
  onSubmit: (parameters: any) => void;
}

/*
 * A component to be placed into a Dialog to gather all necessary parameters from the user for executing the given action.
 */
export default function SirenActionParametersDialogBody({
  action,
  onClose,
  onSubmit
}: Props) {
  const method = action.method || "GET";
  const title = action.title || action.name;

  const [actionPayload, dispatch] = useReducer(actionPayloadReducer, {});

  const formFields = action.fields
    ? action.fields.map(field =>
        renderField(field, actionPayload[field.name] || "", newValue =>
          dispatch({
            type: "onFieldChanged",
            payload: {
              fieldName: field.name,
              newValue
            }
          })
        )
      )
    : [];

  return (
    <React.Fragment>
      <Title method={method} title={title} />
      <form
        onSubmit={event => {
          event.preventDefault();
          onSubmit(actionPayload);
        }}
      >
        <DialogContent>
          {formFields && !formFields.every(elem => elem === null) && (
            <React.Fragment>
              <DialogContentText>Please fill in the form.</DialogContentText>
              {formFields}
            </React.Fragment>
          )}
        </DialogContent>
        <DialogActions>
          <Button
            type={"button"}
            variant={"contained"}
            color={"secondary"}
            onClick={onClose}
            data-cy={"close-button"}
          >
            Cancel
          </Button>
          <Button
            type={"submit"}
            variant={"contained"}
            color={"primary"}
            data-cy={`ok-button`}
          >
            Confirm
          </Button>
        </DialogActions>
      </form>
    </React.Fragment>
  );
}

function renderField(
  field: Field,
  value: any,
  onChange: (newValue: any) => void
) {
  /* Do not display a reason field for declining until we're ready to
     support that */
  if (field.name === "reason") {
    return null;
  }

  if (field.class.includes("ethereum") && field.class.includes("address")) {
    return (
      <EthereumAddressTextField
        data-cy={"action-text-field"}
        required={true}
        key={field.name}
        label={field.title || field.name}
        onAddress={onChange}
        onChange={event => onChange(event.target.value)}
        value={value}
      />
    );
  }

  if (field.class.includes("bitcoin") && field.class.includes("address")) {
    return (
      <BitcoinNanoLedgerAddressTextField
        data-cy={"action-text-field"}
        required={true}
        key={field.name}
        label={field.title || field.name}
        onAddress={onChange}
        onChange={event => onChange(event.target.value)}
        value={value}
      />
    );
  }

  return (
    <TextField
      data-cy={"action-text-field"}
      required={true}
      key={field.name}
      label={field.title || field.name}
      type={field.type}
      value={value}
      onChange={event => onChange(event.target.value)}
    />
  );
}

interface ActionPayload {
  [name: string]: any;
}

// tslint:disable-next-line:interface-over-type-literal
type ReducerAction = {
  type: "onFieldChanged";
  payload: {
    fieldName: string;
    newValue: any;
  };
};

function actionPayloadReducer(state: ActionPayload, action: ReducerAction) {
  switch (action.type) {
    case "onFieldChanged": {
      return {
        ...state,
        [action.payload.fieldName]: action.payload.newValue
      };
    }
  }

  return {};
}

interface DisplayProps {
  method: string;
  title: string;
}

function Title({
  method,
  title,
  ...dialogTitleProps
}: DisplayProps & DialogTitleProps) {
  switch (method) {
    case "GET":
      return (
        <DialogTitle
          {...dialogTitleProps}
        >{`Fetch the '${title}' action`}</DialogTitle>
      );
    default:
      return (
        <DialogTitle
          {...dialogTitleProps}
        >{`Execute the '${title}' action`}</DialogTitle>
      );
  }
}
