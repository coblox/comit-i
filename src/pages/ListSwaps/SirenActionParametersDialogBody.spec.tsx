import { ThemeProvider } from "@material-ui/styles";
import React from "react";
import renderer from "react-test-renderer";
import appTheme from "../../theme";
import SirenActionParametersDialogBody from "./SirenActionParametersDialogBody";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={appTheme}>
        <SirenActionParametersDialogBody
          action={{
            title: "Redeem",
            method: "GET",
            href: "/abcd/redeem",
            name: "redeem",
            fields: [
              {
                title: "Beta ledger redeem identity",
                type: "text",
                name: "beta_ledger_redeem_identity",
                class: ["bitcoin", "address"]
              },
              {
                title: "Fee per byte",
                type: "number",
                name: "feePerByte",
                class: ["bitcoin", "feePerByte"]
              }
            ]
          }}
          onClose={() => ({})}
          onSubmit={() => ({})}
        />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
