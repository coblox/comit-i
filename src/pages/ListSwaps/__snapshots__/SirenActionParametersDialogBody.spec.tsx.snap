// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="MuiDialogTitle-root"
  >
    <h6
      className="MuiTypography-root MuiTypography-h6"
    >
      To fetch the 'Redeem' action, please fill these fields.
    </h6>
  </div>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiDialogContent-root"
    >
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-textField-34"
        data-cy="action-text-field"
      >
        <label
          className="MuiFormLabel-root Mui-required Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Beta ledger redeem identity
          <span
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-textField-34"
        data-cy="action-text-field"
      >
        <label
          className="MuiFormLabel-root Mui-required Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
        >
          Fee per byte
          <span
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="number"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiDialogActions-root MuiDialogActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-containedPrimary MuiButton-contained"
        data-cy="redeem-button"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Fetch Redeem
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-containedSecondary MuiButton-contained"
        data-cy="close-button"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Close
        </span>
      </button>
    </div>
  </form>,
]
`;
